ocaml/typing/typedtree.mli

```diff
@@ -271,7 +273,8 @@ and expression_desc =
           saturated with n arguments.
       *)
   | Texp_apply of
-      expression * (arg_label * apply_arg) list * apply_position * Mode.Locality.l
+      expression * (arg_label * apply_arg) list * apply_position *
+        Mode.Locality.l * Assume_info.t
```

Comment:
Instead of propagating two separate types `Builtin_attributes.check_attribute` and `Assume_info.t`  here, we can extend `Assume_info.t` to propagate `Check` info not only `Assume`. Is there a reason not to in this PR? I'm guessing that a subsequent PR will convert the `zero_alloc` field of  `Text_function` from `Builtin_attributes.checkattribute` to `Assume_info.t` to check inclusion anyway.

Reply:
On a separate note, I would find this clearer if `Texp_apply` carried `Assume_info.t option` instead. It seems like `Assume_info.t` can sneakily represent "None" &mdash; that seems kind of odd, and I'd like to get rid of `Assume_info.is_none` if that's possible.

Reply:
What is the advantage of your proposal? Does `Texp_apply` need to know what `Assume_info.t` is and whether it is `None` or not? `Assume_info.is_none` is used by `Debuginfo.is_none`. Explicitly tracking `option` here would make it harder to change the representation.

Reply:
To Greta's suggestion: The reason some places of the typed tree have `check_attribute` and other places have `assume_info` is just that these types are what is needed by the existing support for `zero_alloc` after typechecking.  For example, lambda's `function_attribute` type wants a `check_attribute`, and the support for putting assumes on scopes wants a `Assume_info.t`.  We could definitely consider trying to clean up those interfaces to agree, but that starts to touch later parts of the compiler, and I propose putting that off for a future PR that isn't focused on the front-end.

In general I already feel bad that this PR does some amount of moving files around and some amount of refactoring - it makes the history nicer when moves and changes happen in separate commits.  I have left a CR in the `Assume_info` module (now part of `Zero_alloc_utils` as requested, see below) to revisit this.

To Nick's suggestion: For the same reason, I don't really want to make changes to the `Assume_info.t` type in this PR, which is meant to be about front-end refactoring.  Such changes would reach further into other parts of the compiler.  I have left a comment in `typed_tree.mli` explaining this field - maybe that, along with the CR mentioned above, is sufficient for now?

Reply:
Happy to defer!

Reply:
sounds good, thanks!